apps/prairielearn/src/pages/studentAssessmentExam/studentAssessmentExam.js

```diff
@@ -1,201 +1,211 @@
         }
-      }
-    );
-  } else if (req.body.__action === 'create_group') {
-    groupAssessmentHelper.createGroup(
-      req.body.groupName,
-      res.locals.assessment.id,
-      res.locals.user.user_id,
-      res.locals.authn_user.user_id,
-      function (err, succeeded, uniqueGroupName, invalidGroupName, permissions) {
-        if (ERR(err, next)) return;
-        if (succeeded) {
-          res.redirect(req.originalUrl);
-        } else {
-          if (invalidGroupName) {
-            res.locals.invalidGroupName = true;
+      );
+    } else if (req.body.__action === 'join_group') {
+      groupAssessmentHelper.joinGroup(
+        req.body.join_code,
+        res.locals.assessment.id,
+        res.locals.user.user_id,
+        res.locals.authn_user.user_id,
+        function (err, succeeded, groupConfig) {
+          if (ERR(err, next)) return err;
+          if (succeeded) {
+            res.redirect(req.originalUrl);
           } else {
-            res.locals.uniqueGroupName = uniqueGroupName;
+            res.locals.groupConfig = groupConfig;
+            res.locals.groupSize = 0;
+            res.locals.used_join_code = req.body.join_code;
+            res.locals.notInGroup = true;
+            res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
           }
-          res.locals.permissions = permissions;
-          res.locals.groupsize = 0;
-          res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
         }
-      }
-    );
-  } else if (req.body.__action === 'leave_group') {
-    groupAssessmentHelper.leaveGroup(
-      res.locals.assessment.id,
-      res.locals.user.user_id,
-      res.locals.authn_user.user_id,
-      function (err) {
-        if (ERR(err, next)) return;
-        res.redirect(req.originalUrl);
-      }
-    );
-  } else {
-    return next(
-      error.make(400, 'unknown __action', {
-        locals: res.locals,
-        body: req.body,
-      })
-    );
-  }
-});
+      );
+    } else if (req.body.__action === 'create_group') {
+      groupAssessmentHelper.createGroup(
+        req.body.groupName,
+        res.locals.assessment.id,
+        res.locals.user.user_id,
+        res.locals.authn_user.user_id,
+        function (err, succeeded, uniqueGroupName, invalidGroupName, permissions) {
+          if (ERR(err, next)) return;
+          if (succeeded) {
+            res.redirect(req.originalUrl);
+          } else {
+            if (invalidGroupName) {
+              res.locals.invalidGroupName = true;
+            } else {
+              res.locals.uniqueGroupName = uniqueGroupName;
+            }
+            res.locals.permissions = permissions;
+            res.locals.groupsize = 0;
+            res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+          }
+        }
+      );
+    } else if (req.body.__action === 'update_group_roles') {
+      await groupAssessmentHelper.updateGroupRoles(
```

Comment:
Does this check for `can_assign_roles_at_start` permissions?

Reply:
No, not explicitly in this endpoint. When rendering the page, we only display the form to update group roles to the user(s) with the `can_assign_roles_at_start` permission, but we didn't write anything in the endpoint to explicitly check if the user who sent the request has that permission.

Reply:
We should _never_ rely on "I didn't show the form to the user" as a form of security. It's always important to check on the backend that the user is actually allowed to perform the action they're taking!

Reply:
We'll write a fix for that here and for homework assessments as well! Thanks for pointing it out!